server/services/v1/auth/gotrue.go

```diff
@@ -0,0 +1,601 @@
+		} else {
+			// expired entry, delete it
+			err := deleteApplicationMetadata(ctx, defaultNamespaceId, clientId, metadataKey, g.userStore, g.txMgr)
+			if err != nil {
+				return "", 0, err
+			}
+		}
+	}
+	// make external call
+	getTokenUrl := fmt.Sprintf("%s/token?grant_type=password", g.AuthConfig.GotrueURL)
+
+	payloadValues := url.Values{}
+	payloadValues.Set("username", clientId)
+	payloadValues.Set("password", clientSecret)
+
+	client := &http.Client{}
+
+	getTokenReq, err := http.NewRequestWithContext(ctx, http.MethodPost, getTokenUrl, strings.NewReader(payloadValues.Encode()))
+	if err != nil {
+		log.Err(err).Msg("Failed to call to get token")
+		return "", 0, err
+	}
+	getTokenReq.Header.Add(GotrueAudHeaderKey, g.AuthConfig.Audience)
+	getTokenReq.Header.Add("Content-Type", "application/x-www-form-urlencoded")
+
+	getTokenRes, err := ctxhttp.Do(ctx, client, getTokenReq)
+	if err != nil {
+		log.Err(err).Msg("Failed to call to get token")
+		return "", 0, err
+	}
+	defer getTokenRes.Body.Close()
+
+	getTokenResBody, err := io.ReadAll(getTokenRes.Body)
+	if err != nil {
+		log.Err(err).Msg("Failed to read getTokenRes body")
+		return "", 0, err
+	}
+
+	if getTokenRes.StatusCode != http.StatusOK {
+		log.Error().Int("status", getTokenRes.StatusCode).Msg("Non OK status received to get access token")
+		return "", 0, errors.Internal("Non OK status code received from gotrue")
+	}
+
+	// parse JSON response
+	var getTokenJsonMap map[string]interface{}
+	err = json.Unmarshal(getTokenResBody, &getTokenJsonMap)
+	if err != nil {
+		log.Err(err).Msg("Failed to deserialize response into JSON")
+		return "", 0, err
+	}
+
+	return getTokenJsonMap["access_token"].(string), int32(getTokenJsonMap["expires_in"].(float64)), nil
+}
+
+func getGotrueAdminAccessToken(ctx context.Context, g *gotrue) (string, int32, error) {
+	return getAccessTokenUsingClientCredentialsGotrue(ctx, g.AuthConfig.GotrueAdminUsername, g.AuthConfig.GotrueAdminPassword, g)
+}
+
+var (
+	idChars     = []rune("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ")
+	secretChars = []rune("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789!@#$%^&*()_+-=~<>;':")
+)
+
+func generateClientId(g *gotrue) string {
+	clientIdLength := g.AuthConfig.GotrueClientIdLength
+	if clientIdLength == 0 {
+		clientIdLength = 30 // default length
+	}
+	b := make([]rune, clientIdLength)
+	for i := range b {
+		b[i] = idChars[generateRandomInt(len(idChars))]
+	}
+	return string(b)
+}
+
+func generateClientSecret(g *gotrue) string {
+	clientSecretLength := g.AuthConfig.GotrueClientSecretLength
+	if clientSecretLength == 0 {
+		clientSecretLength = 50 // default length
+	}
+	b := make([]rune, clientSecretLength)
+	for i := range b {
+		b[i] = secretChars[generateRandomInt(len(secretChars))]
+	}
+	return string(b)
+}
+
+func generateRandomInt(max int) int {
+	var bytes [8]byte
+	_, err := rand.Read(bytes[:])
+	if err != nil {
+		panic(err.Error())
```

Comment:
let's not panic as this will be in online request flow.

Reply:
logged warning.